File: python/makolang/backends/moji.py
import json
from typing import NamedTuple, Optional

import requests

from .common import UA, HTTPRequest, random_uuid, UpstreamAPIError

UNION_API = "https://api.mojidict.com/parse/functions/union-api"
TTS_API = "https://api.mojidict.com/parse/functions/tts-fetch"

SEARCH_FUNCTION_NAME = "search-all"
SEARCH_QUESTION_FUNCTION_NAME = "mojitest-examV2-searchQuestion-v2"
DECONJUGATE_FUNCTION_NAME = "deconjugateWithKeyWord" # 动词变形还原

class MojiWordSearchResult(NamedTuple):
    target_id: str
    title: str
    summary: str
    """单词概要。"""
    summary_b: Optional[str]
    """
    单词概要。
    
    此字段与 `summary` 的区别在于，当单词为动词时，
    `summary` 用一段、五段、サ变、カ变表示动词类型，
    `summary_b` 用一类、二类、三类表示动词类型。
    """

class MojiWordDefinition(NamedTuple):
    pass

def search(word: str) -> list[MojiWordSearchResult]:
    """
    搜索单词。

    :param word: 单词
    :return: 单词信息
    """
    payload = {
        "functions": [
            {
                "name": SEARCH_FUNCTION_NAME,
                "params": {
                    "text": word,
                    # 106 语法、102 单词、103 未知
                    "types": [102, 106, 103]
                }
            }
        ],
        "_ClientVersion": "js3.4.1",
        "_ApplicationId": "E62VyFVLMiW7kvbtVq3p",
        "g_os": "PCWeb",
        "g_ver": "v4.8.9.20241014",
        "_InstallationId": random_uuid(),
    }

    headers = {
        'pragma': 'no-cache',
        'priority': 'u=1, i',
        'User-Agent': UA,
        'content-type': 'text/plain'
    }

    response = requests.request(
        "POST",
        UNION_API,
        headers=headers,
        data=json.dumps(payload)
    )
    
    ret = response.json()
    if ret['result']['code'] != 200:
        UpstreamAPIError(ret['code'], None, response.text)
    
    # result.results["search-all"].result.word.searchResult
    results = []
    for result in ret['result']['results']['search-all']['result']['word']['searchResult']:
        results.append(MojiWordSearchResult(
            result['targetId'],
            result['title'],
            result['excerpt'],
            result.get('excerptB', None)
        ))
    return results

def pronounce(target_id: str) -> HTTPRequest:
    """
    获取单词发音。
    
    :param target_id: 单词 ID。由 `search` 函数返回。
    :return: 发音音频 URL
    """
    payload = {
        "tarId": target_id,
        "tarType": 102,
        "voiceId": "f002", # TODO: 查看此字段的含义
        "_ClientVersion": "js3.4.1", 
        "_ApplicationId": "E62VyFVLMiW7kvbtVq3p",
        "g_os": "PCWeb",
        "g_ver": "v4.8.9.20241014",
        "_InstallationId": random_uuid()
    }

    headers = {
        'pragma': 'no-cache',
        'priority': 'u=1, i',
        'User-Agent': UA,
        'content-type': 'text/plain'
    }
    
    response = requests.request(
        "POST",
        TTS_API,
        headers=headers,
        data=json.dumps(payload)
    )
    ret = response.json()
    if ret['result']['code'] != 200:
        UpstreamAPIError(ret['code'], None, response.text)

    return HTTPRequest(
        method="GET",
        url=ret['result']['result']['url'],
        headers={
            "User-Agent": UA,
        },
        payload=None
    )

def pronounce_x(
    kanji: str,
    kana: str,
    allow_fallback: bool = False
) -> HTTPRequest|None:
    """
    获取单词发音。
    
    :param kanji: 单词
    :param kana: 假名
    :param allow_fallback: 当未找到完全匹配的单词与假名时，是否允许直接返回第一个搜索结果
    """
    results = search(kanji)
    target_id = None
    for result in results:
        title = result.title
        parts = title.split('|')
        if len(parts) == 2:
            reading_part = parts[1].strip()
            reading = reading_part.split(' ')[0].strip()
            if reading == kana:
                target_id = result.target_id
                break

    if not target_id and allow_fallback:
        target_id = results[0].target_id

    if target_id:
        return pronounce(target_id)
    else:
        return None
    

__exports = [
    MojiWordSearchResult,
    MojiWordDefinition,
    search,
    pronounce,
    pronounce_x
]

if __name__ == "__main__":
    from pprint import pprint as print
    from makolang.utility import play_sound
    results = (search("見る"))
    result = results[0]
    tts = pronounce(result.target_id)
    print(tts)
    # tts.save("pronounce.mp3")
    play_sound(tts)

File: php/makolang/backends/moji.php
<?php
require_once dirname(__FILE__).'/../utility/cache.php';
require_once 'common.php';

define('UNION_API', 'https://api.mojidict.com/parse/functions/union-api');
define('TTS_API', 'https://api.mojidict.com/parse/functions/tts-fetch');

define('SEARCH_FUNCTION_NAME', 'search-all');
define('SEARCH_QUESTION_FUNCTION_NAME', 'mojitest-examV2-searchQuestion-v2');
define('DECONJUGATE_FUNCTION_NAME', 'deconjugateWithKeyWord'); // 动词变形还原

class MojiWordSearchResult {
    public $target_id;
    public $title;
    public $summary;
    public $summary_b;

    /**
     * MojiWordSearchResult constructor.
     * 
     * @param string $target_id
     * @param string $title
     * @param string $summary
     * @param string|null $summary_b
     */
    public function __construct($target_id, $title, $summary, $summary_b = null) {
        $this->target_id = $target_id;
        $this->title = $title;
        $this->summary = $summary;
        $this->summary_b = $summary_b;
    }
}

class MojiWordDefinition {
    // 定义类
}

/**
 * 搜索单词。
 * 
 * @param string $word 单词
 * @return array 单词信息
 */
function search($word) {
    $payload = array(
        "functions" => array(
            array(
                "name" => SEARCH_FUNCTION_NAME,
                "params" => array(
                    "text" => $word,
                    "types" => array(102, 106, 103)
                )
            )
        ),
        "_ClientVersion" => "js3.4.1",
        "_ApplicationId" => "E62VyFVLMiW7kvbtVq3p",
        "g_os" => "PCWeb",
        "g_ver" => "v4.8.9.20241014",
        "_InstallationId" => random_uuid(),
    );

    $headers = array(
        'pragma: no-cache',
        'priority: u=1, i',
        'User-Agent: ' . UA,
        'content-type: text/plain'
    );

    $options = array(
        'http' => array(
            'method' => 'POST',
            'header' => implode("\r\n", $headers),
            'content' => json_encode($payload)
        )
    );
    $context = stream_context_create($options);
    $response = cached_file_get_contents(UNION_API, false, $context);
    $ret = json_decode($response, true);

    if ($ret['result']['code'] != 200) {
        throw new UpstreamAPIError($ret['code'], null, $response);
    }

    $results = array();
    foreach ($ret['result']['results']['search-all']['result']['word']['searchResult'] as $result) {
        $results[] = new MojiWordSearchResult(
            $result['targetId'],
            $result['title'],
            $result['excerpt'],
            isset($result['excerptB']) ? $result['excerptB'] : null
        );
    }
    return $results;
}

/**
 * 获取单词发音。
 * 
 * @param string $target_id 单词 ID。由 `search` 函数返回。
 * @return HTTPRequest 发音音频 URL
 */
function pronounce($target_id) {
    $payload = array(
        "tarId" => $target_id,
        "tarType" => 102,
        "voiceId" => "f002",
        "_ClientVersion" => "js3.4.1",
        "_ApplicationId" => "E62VyFVLMiW7kvbtVq3p",
        "g_os" => "PCWeb",
        "g_ver" => "v4.8.9.20241014",
        "_InstallationId" => random_uuid()
    );

    $headers = array(
        'pragma: no-cache',
        'priority: u=1, i',
        'User-Agent: ' . UA,
        'content-type: text/plain'
    );

    $options = array(
        'http' => array(
            'method' => 'POST',
            'header' => implode("\r\n", $headers),
            'content' => json_encode($payload)
        )
    );
    $context = stream_context_create($options);
    $response = cached_file_get_contents(TTS_API, false, $context);
    $ret = json_decode($response, true);

    if ($ret['result']['code'] != 200) {
        throw new UpstreamAPIError($ret['code'], null, $response);
    }

    return new HTTPRequest(
        $ret['result']['result']['url'],
        'GET',
        array('User-Agent' => UA),
        null
    );
}

/**
 * 获取单词发音，支持直接通过汉字和假名获取发音。
 * 
 * @param string $kanji 汉字
 * @param string|null $katakana 假名
 * @param bool $allow_fallback 是否允许回退到第一个搜索结果
 * @return HTTPRequest 发音音频 URL
 * @throws Exception
 */
function pronounce_x($kanji, $katakana = null, $allow_fallback = false) {
    // TODO !important 支持 katakana 与 kanji 二选一
    $results = search($kanji);
    $target_id = null;

    if ($katakana) {
        foreach ($results as $result) {
            $title = $result->title;
            $parts = explode('|', $title);
            if (count($parts) == 2) {
                $reading_part = trim($parts[1]);
                $reading = trim(explode(' ', $reading_part)[0]);
                if ($reading === $katakana) {
                    $target_id = $result->target_id;
                    break;
                }
            }
        }
    }

    if (!$target_id && $allow_fallback) {
        $target_id = $results[0]->target_id;
    }

    if ($target_id) {
        return pronounce($target_id);
    } else {
        throw new Exception('No matching word found');
    }
}

// 注册暴露 API
register_backend('moji', 'word_search', function($lang, $word, $params) {
    if ($lang !== 'ja') {
        throw new UnSupportedLanguageException($lang);
    }
    return search($word);
});

register_backend('moji', 'word_pronounce', function($lang, $word, $params) {
    if ($lang !== 'ja') {
        throw new UnSupportedLanguageException($lang);
    }
    return pronounce_x($word, isset($params['katakana']) ? $params['katakana'] : null, isset($params['allow_fallback']) ? $params['allow_fallback'] : false);
});

// 导出类和函数
$__exports = array(
    'MojiWordSearchResult',
    'MojiWordDefinition',
    'search',
    'pronounce'
);
?>